File: 1930. Unique Length-3 Palindromic Subsequences/main.py
from collections import Counter


class Solution:
    def countPalindromicSubsequence(self, s: str) -> int:
        res = set() # (mid_c, outer_c)
        left = set()
        right = Counter(s)

        for m in s:
            right[m] -= 1
            for c in left:
                if right[c] > 0:
                    res.add((m, c))
            left.add(m)

        return len(res)

File: 1930. Unique Length-3 Palindromic Subsequences/main.go
package main

func countPalindromicSubsequence(s string) int {
	res := make(map[[2]rune]bool)
	left := make(map[rune]bool)
	right := make(map[rune]int)

	for _, c := range s {
		right[c]++
	}

	for _, m := range s {
		right[m]--

		for c := range left {
			if right[c] > 0 {
				res[[2]rune{m, c}] = true
			}
		}

		left[m] = true
	}

	return len(res)
}
